### PR TITLE
Fix NULL Checks and Error Handling

### PR DESCRIPTION
**Pull request by songlingatpan**
This PR primarily addresses potential NULL pointer dereference issues, enhances error handling, and resolves possible memory leaks in failure scenarios.

There are no new features or behavior changes introduced in this PR; the modifications are solely focused on improving stability.

### CODE REVIEW

**Comment by songlingatpan** [comment]
@baentsch This PR is ready for review. Thanks

File: oqsprov/oqs_kmgmt.c [hunk 1]
```diff
@@ -139,13 +142,6 @@ static int oqsx_match(const void *keydata1, const void *keydata2,
     }
 
 #ifdef NOPUBKEY_IN_PRIVKEY
-    /* Now this is a "leap of faith" logic: If a public-only PKEY and a
```
Comment (baentsch) [inline oqsprov/oqs_kmgmt.c:142]: Why do you suggest dropping this comment @songlingatpan ?

File: oqsprov/oqs_kmgmt.c [hunk 2]
```diff
@@ -178,15 +174,13 @@ static int oqsx_match(const void *keydata1, const void *keydata2,
             (key1->pubkey != NULL && key2->pubkey == NULL) ||
             ((key1->tls_name != NULL && key2->tls_name != NULL) &&
              strcmp(key1->tls_name, key2->tls_name))) {
-            // special case now: If domain parameter matching
```
Comment (baentsch) [inline oqsprov/oqs_kmgmt.c:181]: Same here -- why remove comment?

File: oqsprov/oqs_kmgmt.c [hunk 3]
```diff
@@ -264,17 +258,12 @@ int oqsx_key_to_params(const OQSX_KEY *key, OSSL_PARAM_BLD *tmpl,
...
     OQSX_KEY *key = keydata;
-    OSSL_PARAM_BLD *tmpl;
+    OSSL_PARAM_BLD *tmpl = NULL;
     OSSL_PARAM *params = NULL;
-    OSSL_PARAM *p;
     int ok = 1;
 
     OQS_KM_PRINTF("OQSKEYMGMT: export called\n");
 
-    /*
```
Comment (baentsch) [inline oqsprov/oqs_kmgmt.c:274]: Why removed?

File: oqsprov/oqs_kmgmt.c [hunk 4]
```diff
@@ -293,15 +282,18 @@ static int oqsx_export(void *keydata, int selection, OSSL_CALLBACK *param_cb,
         ok = ok && oqsx_key_to_params(key, tmpl, NULL, include_private);
     }
 
-    params = OSSL_PARAM_BLD_to_param(tmpl);
-    if (params == NULL) {
-        ok = 0;
-        goto err;
+    if (ok) {
```
Comment (baentsch) [inline oqsprov/oqs_kmgmt.c:285]: Why not also asserting `tmpl!=NULL`?

File: oqsprov/oqs_kmgmt.c [hunk 5]
```diff
@@ -642,27 +700,30 @@ static int oqsx_gen_set_params(void *genctx, const OSSL_PARAM params[]) {
...
-        OPENSSL_free(gctx->propq);
-        gctx->propq = OPENSSL_strdup(p->data);
-        if (gctx->propq == NULL)
+        char *new_propq = OPENSSL_strdup(p->data);
+        if (new_propq == NULL)
             return 0;
+        OPENSSL_free(gctx->propq);
+        gctx->propq = new_propq;
     }
-    // not passing in params is no error; subsequent operations may fail,
```
Comment (baentsch) [inline oqsprov/oqs_kmgmt.c:664]: why deleted?

File: oqsprov/oqsprov_keys.c [hunk 6]
```diff
@@ -403,27 +423,44 @@ EVP_PKEY *setECParams(EVP_PKEY *eck, int nid) {
...
+
+    if (eck == NULL) {
+        return NULL;
+    }
+
     switch (nid) {
     case NID_X9_62_prime256v1:
         params = p256params;
-        return d2i_KeyParams(EVP_PKEY_EC, &eck, &params, sizeof(p256params));
+        params_len = sizeof(p256params);
```
Comment (baentsch) [inline oqsprov/oqsprov_keys.c:436]: This seems like a semantics change: Can you please explain the rationale, @songlingatpan ?

File: oqsprov/oqsprov_keys.c [hunk 7]
```diff
@@ -403,27 +423,44 @@ EVP_PKEY *setECParams(EVP_PKEY *eck, int nid) {
...
         params = bp384params;
-        return d2i_KeyParams(EVP_PKEY_EC, &eck, &params, sizeof(bp384params));
+        params_len = sizeof(bp384params);
+        break;
     default:
         return NULL;
     }
-}
 
+    result = d2i_KeyParams(EVP_PKEY_EC, &eck, &params, params_len);
```
Comment (baentsch) [inline oqsprov/oqsprov_keys.c:458]: Ahh -- as it's pulled down here. OK, then. Understood.

File: oqsprov/oqsprov_keys.c [hunk 8]
```diff
@@ -566,269 +629,334 @@ static const int oqshybkem_init_ecx(char *tls_name, OQSX_EVP_CTX *evp_ctx) {
...
+
+free_key_param:
+    EVP_PKEY_free(evp_ctx->keyParam);
+    evp_ctx->keyParam = NULL;
 
 err_init_ecx:
     return ret;
 }
 
-/* Re-create OQSX_KEY from encoding(s): Same end-state as after ken-gen */
```
Comment (baentsch) [inline oqsprov/oqsprov_keys.c:578]: Please bring back comment as it aids understanding.

**Review by baentsch — CHANGES_REQUESTED** [review]
Thanks for the substantial amount of work, @songlingatpan ! This clearly improves the code resilience. Due to the size of the PR I had to stop in the middle of the review but would like to ask you to bring back the comments deleted (or explain why you deleted them) as well as provide an explanation what you did to the composite logic so I can better follow all changes there. Thanks!

**Comment by songlingatpan** [comment]
> Thanks for the substantial amount of work, @songlingatpan ! This clearly improves the code resilience. Due to the size of the PR I had to stop in the middle of the review but would like to ask you to bring back the comments deleted (or explain why you deleted them) as well as provide an explanation what you did to the composite logic so I can better follow all changes there. Thanks!

Sorry @baentsch. I will add comments back.
The oqsx_key_op function was quite large, so I refactored it into smaller, more functions such as allocate_and_copy_keymaterial, oqsx_key_process_composite_keys, and process_private_key. However, the logic remains unchanged.

Thanks a lot again for your incredible work on the OQS provider.

File: oqsprov/oqs_kem.c [hunk 9]
```diff
@@ -67,19 +67,23 @@ static void oqs_kem_freectx(void *vpkemctx) {
     PROV_OQSKEM_CTX *pkemctx = (PROV_OQSKEM_CTX *)vpkemctx;
 
     OQS_KEM_PRINTF("OQS KEM provider called: freectx\n");
-    oqsx_key_free(pkemctx->kem);
-    OPENSSL_free(pkemctx);
+    if (pkemctx != NULL) {
+        oqsx_key_free(pkemctx->kem);
+        OPENSSL_free(pkemctx);
```
Comment (SWilson4) [inline oqsprov/oqs_kem.c:72]: Same comment as below: `OPENSSL_free` handles NULL gracefully, and imo it's cleaner to not perform the check.

**Comment by baentsch** [comment]
> > this makes it easier to reason about code correctness and safety.

> It would also be good to hear @baentsch's perspective on this as the primary developer and maintainer of the codebase.

I agree -- the ...free.... functions typically handle NULL objects gracefully. And I also agree, too many safety checks make the code hard to read.

But then again, there's _lots of other code_ that's harder to read/understand/maintain than these added checks.

Maybe using macros might be a compromise?

File: oqsprov/oqs_kem.c (hunk 9 above)
Comment (songlingatpan) [reply oqsprov/oqs_kem.c:72]: The NULL check is for prevent from NULL pointer for oqsx_key_free(pkemctx->kem), not for OPENSSL_free/free().

File: oqsprov/oqs_kem.c [hunk 10]
```diff
@@ -67,19 +67,23 @@ static void oqs_kem_freectx(void *vpkemctx) {
...
                                      int operation) {
     PROV_OQSKEM_CTX *pkemctx = (PROV_OQSKEM_CTX *)vpkemctx;
 
     OQS_KEM_PRINTF3("OQS KEM provider called: _init : New: %p; old: %p \n",
-                    vkem, pkemctx->kem);
+                    vkem, pkemctx ? pkemctx->kem : NULL);
     if (pkemctx == NULL || vkem == NULL || !oqsx_key_up_ref(vkem))
         return 0;
-    oqsx_key_free(pkemctx->kem);
+    if (pkemctx->kem != NULL) {
```
[the remark this answers is not in the document]
Comment (songlingatpan) [reply oqsprov/oqs_kem.c:84]: Thanks for the comments. 
We see quite some crashes due to dereference NULL pointer in our load testing. That is why we aggressively add NULL checks.
I will remove unnecessary NULL checks.

**Comment by songlingatpan** [comment]
> > > this makes it easier to reason about code correctness and safety.
> 
> > It would also be good to hear @baentsch's perspective on this as the primary developer and maintainer of the codebase.
> 
> I agree -- the ...free.... functions typically handle NULL objects gracefully. And I also agree, too many safety checks make the code hard to read.
> 
> But then again, there's _lots of other code_ that's harder to read/understand/maintain than these added checks.
> 
> Maybe using macros might be a compromise?

Thanks for the comments. 
We see quite some crashes due to dereference NULL pointer in our load testing. That is why we aggressively add NULL checks.
I will remove unnecessary NULL checks.

**Comment by songlingatpan** [comment]
@baentsch @SWilson4 
When we encountered crashes, we reviewed the related functions and added NULL checks. I agree that, in some cases, we may have been overly cautious with these checks. The reason for this approach is that we were consistently seeing crashes in various places, so we aimed to go through the files one by one and add protective code.

I will go ahead and remove the unnecessary NULL checks.

Thank you again for the excellent work you've done on building the OQS code.


`#0  0x00007ffff67c20f2 in evp_md_init_internal (ctx=0x0, type=0xd4661daf00, params=0x0, impl=0x0) at crypto/evp/digest.c:166

#1  0x00007ffff67c26ee in EVP_DigestInit_ex (ctx=<optimized out>, type=<optimized out>, impl=<optimized out>) at crypto/evp/digest.c:385

#2  0x00007ffff1a9ac63 in do_hash (md=0xd4661daf00, inplen=64, input=0x7fffdfd8cde0 "s\315~jQ\261\224 \032\025\346\243\201\354\340\240\004:*7\020o`e\177\326d\263Z\215\251\253\225E\002\361\340L\305\b\334\nD;\365d5\004\005\066\302\060\264\024\060<\n\331", output=0x7fffdfd8ce20 "") at ../src/common/sha3/ossl_sha3.c:22

#3  OQS_SHA3_sha3_512 (output=0x7fffdfd8ce20 "", input=0x7fffdfd8cde0 "s\315~jQ\261\224 \032\025\346\243\201\354\340\240\004:*7\020o`e\177\326d\263Z\215\251\253\225E\002\361\340L\305\b\334\nD;\365d5\004\005\066\302\060\264\024\060<\n\331", inplen=64) at ../src/common/sha3/ossl_sha3.c:110

#4  0x00007ffff1936a32 in pqcrystals_kyber1024_ref_dec (ss=0xd46bc1fa40 "", ct=0xd46970715a "\362,\325\005\240\317_NTC\267F3\233\325\234k\020\034\255\253\277\032\355\244\302^F\v*\312\334\315B\030\232\263\327j-\270\233~\236\352\304\260G'\016\252\217\274/}\275\r\ad\321!e5\305\356/\304\345oB \v\271\206\324\355/\346\211\302\236Z\337\311\230\243\334\370\314\246\354zD4\327{\373\063\315\366\372\347\233\233 \232\031\215<w\321\267/\031<4\021\226\065\360\242\226\240\347\063\267E\241I\301R\213\201\224\364/\354+\027\017\033\227\224Q\"\327\311\233_\300\217\315H\343\370\273\346N\337\226=\r\305c\274\347t\332U\036\331\200\370l\324MvAi\365\356E\005\243\025mLK\277x4W\371i\235t4\211"..., sk=0xd46630c400 "l%G\342

{eD\232\062\363dh@\003V\213\311\244\375\303N4D\236/[\\~\020)\212\240\310\363\200=\222V\a\304su\003\206\237|s~^)\313-\n\315\334v5Q\262\060\200\273i\322\227}
r`

`#0  OQS_SHA3_shake128_inc_squeeze (output=output@entry=0x7fffe1cb4590 "", outlen=outlen@entry=504, state=state@entry=0x7fffe1cb4588) at ../src/common/sha3/ossl_sha3.c:219

#1  0x00007ffff19362d1 in pqcrystals_kyber1024_ref_gen_matrix (a=a@entry=0x7fffe1cb5ff0, seed=seed@entry=0x7fffe1cb7ff0 "3nQ\371\341\231\063\271\067\305\370\240\264\306yTz\373\20

6O\353Z\251\030\001\t\006\255\376\210v!", transposed=transposed@entry=0) at ../src/kem/kyber/pqcrystals-kyber_kyber1024_ref/indcpa.c:179

#2  0x00007ffff1936454 in pqcrystals_kyber1024_ref_indcpa_keypair (pk=pk@entry=0xd46821a200 "", sk=sk@entry=0xd468509d00 "") at ../src/kem/kyber/pqcrystals-kyber_kyber1024_ref/in

dcpa.c:220

#3  0x00007ffff1936885 in pqcrystals_kyber1024_ref_keypair (pk=0xd46821a200 "", sk=0xd468509d00 "") at ../src/kem/kyber/pqcrystals-kyber_kyber1024_ref/kem.c:27

#4  0x00007ffff1f0afa5 in oqsx_key_gen_oqs (gen_kem=1, key=0xd036b634c0) at ../oqsprov/oqsprov_keys.c:1735

#5  oqsx_key_gen_oqs (gen_kem=1, key=0xd036b634c0) at ../oqsprov/oqsprov_keys.c:1732

#6  oqsx_key_gen (key=key@entry=0xd036b634c0) at ../oqsprov/oqsprov_keys.c:1858

#7  0x00007ffff1f0e85d in oqsx_genkey (gctx=0xd467fcc9c0) at ../oqsprov/oqs_kmgmt.c:613

`

File: oqsprov/oqs_kem.c (hunk 9 above)
Comment (SWilson4) [reply oqsprov/oqs_kem.c:72]: Ah yes, I should have been more clear: I meant to write that the `OPENSSL_free` call could be moved outside the conditional block, not that the check itself could be removed entirely. Certainly it's still necessary before the dereference.

**Comment by baentsch** [comment]
@songlingatpan Thanks again for the diligent addition of safeguards to the code. However, I'm not sure I understand your latest comment: What should we do with these line items? Please also re-base to latest main. #524 changed a bit too much impacting other PRs :-(

**Comment by baentsch** [comment]
Closing due to inactivity.

**Pull request closed without merging**